core/store/src/trie/mem/mod.rs

```diff
@@ -1,8 +1,206 @@
+    /// trie nodes. This internally takes care of refcounting.
+    pub fn construct_root<Error>(
+        &mut self,
+        block_height: BlockHeight,
+        f: impl FnOnce(&mut Arena) -> Result<MemTrieNodeId, Error>,
+    ) -> Result<CryptoHash, Error> {
+        let root = f(&mut self.arena)?;
+        let state_root = root.as_ptr(self.arena.memory()).view().node_hash();
+        self.insert_root(state_root, root, block_height);
+        Ok(state_root)
+    }
+
+    fn insert_root(
+        &mut self,
+        state_root: StateRoot,
+        mem_root: MemTrieNodeId,
+        block_height: BlockHeight,
+    ) {
+        assert_ne!(state_root, CryptoHash::default());
+        let heights = self.heights.entry(block_height).or_default();
+        heights.push(state_root);
+        let new_ref = mem_root.add_ref(&mut self.arena);
+        if new_ref == 1 {
+            self.roots.entry(state_root).or_default().push(mem_root);
+        }
+        MEM_TRIE_NUM_ROOTS
+            .with_label_values(&[&self.shard_uid.shard_id.to_string()])
+            .set(self.roots.len() as i64);
+    }
+
+    /// Returns a root node corresponding to the given state root.
+    pub fn get_root<'a>(&'a self, state_root: &CryptoHash) -> Option<MemTrieNodePtr<'a>> {
+        assert_ne!(state_root, &CryptoHash::default());
+        self.roots.get(state_root).map(|ids| ids[0].as_ptr(self.arena.memory()))
+    }
+
+    /// Expires all trie roots corresponding to a height smaller than
+    /// `block_height`. This internally manages refcounts. If a trie root
+    /// is expired but is still used at a higher height, it will still be
+    /// valid until all references to that root expires.
+    pub fn delete_until_height(&mut self, block_height: BlockHeight) {
+        let mut to_delete = vec![];
+        self.heights.retain(|height, state_roots| {
+            if *height < block_height {
+                for state_root in state_roots {
+                    to_delete.push(*state_root)
+                }
+                false
+            } else {
+                true
+            }
+        });
+        for state_root in to_delete {
+            self.delete_root(&state_root);
+        }
+    }
+
+    fn delete_root(&mut self, state_root: &CryptoHash) {
+        if let Some(ids) = self.roots.get_mut(state_root) {
+            let last_id = ids.last().unwrap();
+            let new_ref = last_id.remove_ref(&mut self.arena);
+            if new_ref == 0 {
+                ids.pop();
+                if ids.is_empty() {
+                    self.roots.remove(state_root);
+                }
+            }
+        } else {
+            debug_assert!(false, "Deleting non-existent root: {}", state_root);
+        }
+        MEM_TRIE_NUM_ROOTS
+            .with_label_values(&[&self.shard_uid.shard_id.to_string()])
+            .set(self.roots.len() as i64);
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
So right here I'm explicitly refusing to handle empty tries. I know that in your prototype code you've checked that CryptoHash::default() refers to the empty trie, but for which we return a MemTrieNodeId from an invalid pointer. The invalid pointer part is what I'm less comfortable with. Refcounting also kinda makes no sense because of we're basically not counting it. So I'm hoping that when I add in the code that uses this, I'd be able to special-case the empty trie case.

This test here is testing refcounting for the trie roots, not adding/removing keys; that's not implemented yet.

Reply:
Ah, so having `mem_root: MemTrieNodeId` among `insert_root(...)` args assumes that someone constructed valid `MemTrieNodeId`, got it!
Then I suggest - on returning an error like `assert_ne!(state_root, CryptoHash::default());`, how about adding a message like "detected empty state root which can't correspond to valid MemTrieNodeId"?